File: CoreData/Objective-C_CoreData/Objective-C_CoreData/ViewController.m
//
//  ViewController.m
//  Objective-C_CoreData
//
//  Created by Kxx.xxQ 一生相伴 on 16/2/28.
//  Copyright © 2016年 Asahi_Kuang. All rights reserved.
//

#import "ViewController.h"
#import "AppDelegate.h"
#import <CoreData/CoreData.h>

static NSString *const identifier = @"cellIdentifier";
@interface ViewController ()<UITableViewDataSource, UITableViewDelegate>
@property (weak, nonatomic) IBOutlet UITableView *listTableView;

// 数组 存储core data 实体对象。
@property (strong, nonatomic) NSMutableArray *listItems;
@end

@implementation ViewController

#pragma mark - lazy loading
- (NSMutableArray *)listItems {
    if (!_listItems) {
        _listItems = [NSMutableArray array];
    }
    return _listItems;
}
#pragma mark --

#pragma mark - life cycel
- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    _listTableView.delegate = self;
    _listTableView.dataSource = self;
    _listTableView.tableFooterView = [UIView new];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    // 查询数据
    [self executeDataItem];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}
#pragma mark --

#pragma mark - selector
- (IBAction)addNewItem:(UIBarButtonItem *)sender {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"New Item" message:@"add new item to core data." preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *completeAction = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        //
        
        if (alert.textFields.firstObject.text.length == 0) {
            return;
        }
        [self addNewItems:alert.textFields.firstObject.text];
        
        
    }];
    UIAlertAction *cancel = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        //
        textField.placeholder = @"Input new item here.";
    }];
    [alert addAction:completeAction];
    [alert addAction:cancel];
    [self presentViewController:alert animated:YES completion:nil];
}
#pragma mark --

#pragma mark - data handler
// 获取上下文
- (NSManagedObjectContext *)getContext {
    AppDelegate *delegate = [UIApplication sharedApplication].delegate;
    return delegate.managedObjectContext;
}

// 新增数据 方法
- (void)addNewItems:(NSString *)newItemName {
    
    NSManagedObjectContext *context = [self getContext];
    // 初始化core data 实体
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"ListItem" inManagedObjectContext:context];
    NSManagedObject *item = [[NSManagedObject alloc] initWithEntity:entity insertIntoManagedObjectContext:context];
    // 给实体属性赋值。
    [item setValue:newItemName forKey:@"itemName"];
    
    NSError *__autoreleasing error = nil;
    
    // 执行存储数据
    [context save:&error];
    if (!error) {
        [self.listItems addObject:item];
    }
    else {
        NSLog(@"存储错误！----> %@", [error localizedDescription]);
    }
    [self.listTableView reloadData];
}

// 查询数据方法。
- (void)executeDataItem {
    NSManagedObjectContext *context = [self getContext];
    // 初始化查询请求。
    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:@"ListItem"];
    NSError *__autoreleasing error = nil;
    // 请求的结果。数组
    NSArray *results = [context executeFetchRequest:request error:&error];
    if (!error) {
        // 将请求的core data 数据放进数据源数组。
        [self.listItems addObjectsFromArray:results];
//
    }
    else {
        NSLog(@"查询错误！----> %@", [error localizedDescription]);
    }
    [self.listTableView reloadData];
}

- (void)deleteDataItemAtIndexPath:(NSIndexPath *)indexPath {
    NSManagedObjectContext *context = [self getContext];
    
    // 从上下文中删除对象
    [context deleteObject:self.listItems[indexPath.row]];
    // 移除数组中对应的实体对象。
    [self.listItems removeObjectAtIndex:indexPath.row];

    [self.listTableView reloadData];
}
#pragma mark --

#pragma mark - UITableViewDelegate && dataSource
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
    }
    NSManagedObject *objectItem = self.listItems[indexPath.row];
    cell.textLabel.text = [objectItem valueForKey:@"itemName"];
    return cell;
}

- (NSArray<UITableViewRowAction *> *)tableView:(UITableView *)tableView editActionsForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewRowAction *deleteAction = [UITableViewRowAction rowActionWithStyle:UITableViewRowActionStyleDefault title:@"删除" handler:^(UITableViewRowAction * _Nonnull action, NSIndexPath * _Nonnull indexPath) {
        //
        [self deleteDataItemAtIndexPath:indexPath];
    }];
    return @[deleteAction];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return [self.listItems count];
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 65;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
}
#pragma mark --

@end

File: CoreData/Swift-CoreData/Swift-CoreData/ViewController.swift
//
//  ViewController.swift
//  Swift-CoreData
//
//  Created by Kxx.xxQ 一生相伴 on 16/2/28.
//  Copyright © 2016年 Asahi_Kuang. All rights reserved.
//

import UIKit
import CoreData

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource, UIAlertViewDelegate {

    let identifier: String = "cellIdentifier"
    // 用来存储coreData实体对象
    var listitems: Array<NSManagedObject> = [NSManagedObject]()
    
    @IBOutlet weak var listTableView: UITableView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        listTableView.delegate = self
        listTableView.dataSource = self
        listTableView.tableFooterView = UIView()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        
        coreDataHandler()
        
    }
    // 查询数据
    func coreDataHandler() {
        let appdelegate = UIApplication.sharedApplication().delegate as! AppDelegate
        // 获取上下文
        let context = appdelegate.managedObjectContext
        // 初始化查询请求
        let request = NSFetchRequest(entityName: "ListItem")
        
        do{
            let results = try context.executeFetchRequest(request)
            listitems = results as! [NSManagedObject]
        }catch{
            print("查询失败!")
        }
        listTableView.reloadData()
    }

    
    // 存储数据
    func storeData(data: String) {
        let appdelegate = UIApplication.sharedApplication().delegate as! AppDelegate
        let context = appdelegate.managedObjectContext
        // 从上下文获取实例
        let entity = NSEntityDescription.entityForName("ListItem", inManagedObjectContext: context)
        // 插入新实体
        let item = NSManagedObject(entity: entity!, insertIntoManagedObjectContext: context)
        // 给实体对象属性赋值
        item.setValue(data, forKey: "itemName")
        
        do{
            // 存储数据
            try context.save()
            listitems.append(item)
        }catch{
            print(error)
        }
        listTableView.reloadData()
    }
    
    // 删除数据
    func deleteData(indexPath: NSIndexPath) {
        let appdelegate = UIApplication.sharedApplication().delegate as! AppDelegate
        let context = appdelegate.managedObjectContext
        // 删除实体对象
        context.deleteObject(self.listitems[indexPath.row])
        // 删除数组对应的实体对象
        self.listitems.removeAtIndex(indexPath.row)
        // 重载对应的row
        self.listTableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: .Automatic)
    }
    
    @IBAction func addNewItem(sender: UIBarButtonItem) {
        let addNewItemAlert = UIAlertController(title: "Add New Item", message: "Input the new item you want to add.", preferredStyle: .Alert)
        let completeAction = UIAlertAction(title: "Ok", style: .Default) { (okButton) -> Void in
            //
            if addNewItemAlert.textFields!.first!.text != nil {
                
                self.storeData(addNewItemAlert.textFields!.first!.text!)
            }
        }
        let cancelAction = UIAlertAction(title: "Cancel", style: .Cancel) { (cancelButton) -> Void in
            //
        }
        addNewItemAlert.addTextFieldWithConfigurationHandler { (inputText) -> Void in
            //
            inputText.placeholder = "Input new item name here."
        }
        addNewItemAlert.addAction(completeAction)
        addNewItemAlert.addAction(cancelAction)
        self.presentViewController(addNewItemAlert, animated: true, completion: nil)
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return listitems.count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        var cell = tableView.dequeueReusableCellWithIdentifier(identifier)
        if cell == nil {
            cell = UITableViewCell(style: .Default, reuseIdentifier: identifier)
        }
        let itemEntity = listitems[indexPath.row]
        cell!.textLabel!.text = itemEntity.valueForKey("itemName") as? String
        return cell!
    }

    func tableView(tableView: UITableView, willDisplayCell cell: UITableViewCell, forRowAtIndexPath indexPath: NSIndexPath) {
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        tableView.deselectRowAtIndexPath(indexPath, animated: false)
    }
    
    func tableView(tableView: UITableView, editActionsForRowAtIndexPath indexPath: NSIndexPath) -> [UITableViewRowAction]? {
        let deleteAction = UITableViewRowAction(style: .Default, title: "删除") { (delete, indexPath) -> Void in
            //

            self.deleteData(indexPath)
        }
        
        deleteAction.backgroundColor = UIColor(red: 41/255.0, green: 202/255.0, blue: 184/255.0, alpha: 1.0)
        return [deleteAction]
    }
    
    func tableView(tableView: UITableView, canEditRowAtIndexPath indexPath: NSIndexPath) -> Bool {return true}
    
    func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {return 65}
}

